wagtail/contrib/modeladmin/tests/test_simple_modeladmin.py

```diff
@@ -755,4 +755,5 @@ def test_inspect_view(self):
         # Ensure that the permissions helper caches the model permissions
         # and only does one query per modeladmin
         with self.assertNumQueries(35):
+            Site.get_site_root_paths()
```

Comment:
Any idea why isn't there any  `auth_permission` queries in the failing of the tests?  When I run it locally these are the 35 queries:

```
1. SELECT "django_session"."session_key", "django_session"."session_data", "django_session"."expire_date" FROM "django_session" WHERE ("django_session"."expire_date" > '2021-04-22 13:19:39.526751' AND "django_session"."session_key" = 'ml9w9xm22nux0w04fappxwojouxql3nd') LIMIT 21
2. SELECT "customuser_customuser"."password", "customuser_customuser"."last_login", "customuser_customuser"."is_superuser", "customuser_customuser"."identifier", "customuser_customuser"."username", "customuser_customuser"."email", "customuser_customuser"."is_staff", "customuser_customuser"."is_active", "customuser_customuser"."first_name", "customuser_customuser"."last_name", "customuser_customuser"."country", "customuser_customuser"."attachment" FROM "customuser_customuser" WHERE "customuser_customuser"."identifier" = 1361840001 LIMIT 21
3. SELECT "wagtailusers_userprofile"."id", "wagtailusers_userprofile"."user_id", "wagtailusers_userprofile"."submitted_notifications", "wagtailusers_userprofile"."approved_notifications", "wagtailusers_userprofile"."rejected_notifications", "wagtailusers_userprofile"."preferred_language", "wagtailusers_userprofile"."current_time_zone", "wagtailusers_userprofile"."avatar" FROM "wagtailusers_userprofile" WHERE "wagtailusers_userprofile"."user_id" = 1361840001 LIMIT 21
4. SELECT "auth_permission"."codename" FROM "auth_permission" INNER JOIN "django_content_type" ON ("auth_permission"."content_type_id" = "django_content_type"."id") WHERE ("django_content_type"."app_label" = 'modeladmintest' AND "django_content_type"."model" = 'author') ORDER BY "django_content_type"."app_label" ASC, "django_content_type"."model" ASC, "auth_permission"."codename" ASC
5. SELECT COUNT(*) AS "__count" FROM "modeladmintest_author"
6. SELECT COUNT(*) AS "__count" FROM "modeladmintest_author"
7. SELECT COUNT(*) AS "__count" FROM "modeladmintest_author"
8. SELECT "wagtailcore_locale"."id", "wagtailcore_locale"."language_code" FROM "wagtailcore_locale" WHERE "wagtailcore_locale"."language_code" IN ('en', 'fr') ORDER BY "wagtailcore_locale"."language_code" ASC
9. SELECT "auth_permission"."codename" FROM "auth_permission" INNER JOIN "django_content_type" ON ("auth_permission"."content_type_id" = "django_content_type"."id") WHERE ("django_content_type"."app_label" = 'modeladmintest' AND "django_content_type"."model" = 'book') ORDER BY "django_content_type"."app_label" ASC, "django_content_type"."model" ASC, "auth_permission"."codename" ASC
10. SELECT "auth_permission"."codename" FROM "auth_permission" INNER JOIN "django_content_type" ON ("auth_permission"."content_type_id" = "django_content_type"."id") WHERE ("django_content_type"."app_label" = 'modeladmintest' AND "django_content_type"."model" = 'token') ORDER BY "django_content_type"."app_label" ASC, "django_content_type"."model" ASC, "auth_permission"."codename" ASC
11. SELECT "auth_permission"."codename" FROM "auth_permission" INNER JOIN "django_content_type" ON ("auth_permission"."content_type_id" = "django_content_type"."id") WHERE ("django_content_type"."app_label" = 'modeladmintest' AND "django_content_type"."model" = 'publisher') ORDER BY "django_content_type"."app_label" ASC, "django_content_type"."model" ASC, "auth_permission"."codename" ASC
12. SELECT "auth_permission"."codename" FROM "auth_permission" INNER JOIN "django_content_type" ON ("auth_permission"."content_type_id" = "django_content_type"."id") WHERE ("django_content_type"."app_label" = 'modeladmintest' AND "django_content_type"."model" = 'person') ORDER BY "django_content_type"."app_label" ASC, "django_content_type"."model" ASC, "auth_permission"."codename" ASC
13. SELECT "auth_permission"."codename" FROM "auth_permission" INNER JOIN "django_content_type" ON ("auth_permission"."content_type_id" = "django_content_type"."id") WHERE ("django_content_type"."app_label" = 'modeladmintest' AND "django_content_type"."model" = 'friend') ORDER BY "django_content_type"."app_label" ASC, "django_content_type"."model" ASC, "auth_permission"."codename" ASC
14. SELECT "auth_permission"."codename" FROM "auth_permission" INNER JOIN "django_content_type" ON ("auth_permission"."content_type_id" = "django_content_type"."id") WHERE ("django_content_type"."app_label" = 'modeladmintest' AND "django_content_type"."model" = 'visitor') ORDER BY "django_content_type"."app_label" ASC, "django_content_type"."model" ASC, "auth_permission"."codename" ASC
15. SELECT "auth_permission"."codename" FROM "auth_permission" INNER JOIN "django_content_type" ON ("auth_permission"."content_type_id" = "django_content_type"."id") WHERE ("django_content_type"."app_label" = 'modeladmintest' AND "django_content_type"."model" = 'contributor') ORDER BY "django_content_type"."app_label" ASC, "django_content_type"."model" ASC, "auth_permission"."codename" ASC
16. SELECT "auth_permission"."codename" FROM "auth_permission" INNER JOIN "django_content_type" ON ("auth_permission"."content_type_id" = "django_content_type"."id") WHERE ("django_content_type"."app_label" = 'modeladmintest' AND "django_content_type"."model" = 'relatedlink') ORDER BY "django_content_type"."app_label" ASC, "django_content_type"."model" ASC, "auth_permission"."codename" ASC
17. SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE ("django_content_type"."app_label" IN ('tests') AND "django_content_type"."model" IN ('formpagewithcustomsubmission', 'jadeformpage', 'formpagewithredirect', 'formpage', 'formpagewithcustomsubmissionlistview', 'formpagewithcustomformbuilder'))
18. SELECT (1) AS "a" FROM "wagtailcore_page" WHERE "wagtailcore_page"."content_type_id" IN (28, 83, 65, 81, 75, 59) LIMIT 1
19. SELECT (1) AS "a" FROM "wagtailcore_page" LIMIT 1
20. SELECT (1) AS "a" FROM "wagtailcore_page" WHERE "wagtailcore_page"."depth" = 1 LIMIT 1
21. SELECT "wagtailcore_page"."path" FROM "wagtailcore_page" WHERE "wagtailcore_page"."depth" = 1
22. SELECT "wagtailcore_page"."id", "wagtailcore_page"."path", "wagtailcore_page"."depth", "wagtailcore_page"."numchild", "wagtailcore_page"."translation_key", "wagtailcore_page"."locale_id", "wagtailcore_page"."title", "wagtailcore_page"."draft_title", "wagtailcore_page"."slug", "wagtailcore_page"."content_type_id", "wagtailcore_page"."live", "wagtailcore_page"."has_unpublished_changes", "wagtailcore_page"."url_path", "wagtailcore_page"."owner_id", "wagtailcore_page"."seo_title", "wagtailcore_page"."show_in_menus", "wagtailcore_page"."search_description", "wagtailcore_page"."go_live_at", "wagtailcore_page"."expire_at", "wagtailcore_page"."expired", "wagtailcore_page"."locked", "wagtailcore_page"."locked_at", "wagtailcore_page"."locked_by_id", "wagtailcore_page"."first_published_at", "wagtailcore_page"."last_published_at", "wagtailcore_page"."latest_revision_created_at", "wagtailcore_page"."live_revision_id", "wagtailcore_page"."alias_of_id" FROM "wagtailcore_page" WHERE "wagtailcore_page"."path" = '0001' LIMIT 21
23. SELECT (1) AS "a" FROM "wagtailcore_page" LIMIT 1
24. SELECT (1) AS "a" FROM "wagtailcore_page" LIMIT 1
25. SELECT "modeladmintest_author"."id", "modeladmintest_author"."name", "modeladmintest_author"."date_of_birth" FROM "modeladmintest_author" ORDER BY "modeladmintest_author"."id" DESC LIMIT 5
26. SELECT "modeladmintest_book"."id", "modeladmintest_book"."author_id", "modeladmintest_book"."title", "modeladmintest_book"."cover_image_id" FROM "modeladmintest_book" WHERE "modeladmintest_book"."author_id" = 5 ORDER BY "modeladmintest_book"."id" ASC LIMIT 1
27. SELECT "modeladmintest_book"."id", "modeladmintest_book"."author_id", "modeladmintest_book"."title", "modeladmintest_book"."cover_image_id" FROM "modeladmintest_book" WHERE "modeladmintest_book"."author_id" = 5 ORDER BY "modeladmintest_book"."id" DESC LIMIT 1
28. SELECT "modeladmintest_book"."id", "modeladmintest_book"."author_id", "modeladmintest_book"."title", "modeladmintest_book"."cover_image_id" FROM "modeladmintest_book" WHERE "modeladmintest_book"."author_id" = 4 ORDER BY "modeladmintest_book"."id" ASC LIMIT 1
29. SELECT "modeladmintest_book"."id", "modeladmintest_book"."author_id", "modeladmintest_book"."title", "modeladmintest_book"."cover_image_id" FROM "modeladmintest_book" WHERE "modeladmintest_book"."author_id" = 4 ORDER BY "modeladmintest_book"."id" DESC LIMIT 1
30. SELECT "modeladmintest_book"."id", "modeladmintest_book"."author_id", "modeladmintest_book"."title", "modeladmintest_book"."cover_image_id" FROM "modeladmintest_book" WHERE "modeladmintest_book"."author_id" = 3 ORDER BY "modeladmintest_book"."id" ASC LIMIT 1
31. SELECT "modeladmintest_book"."id", "modeladmintest_book"."author_id", "modeladmintest_book"."title", "modeladmintest_book"."cover_image_id" FROM "modeladmintest_book" WHERE "modeladmintest_book"."author_id" = 3 ORDER BY "modeladmintest_book"."id" DESC LIMIT 1
32. SELECT "modeladmintest_book"."id", "modeladmintest_book"."author_id", "modeladmintest_book"."title", "modeladmintest_book"."cover_image_id" FROM "modeladmintest_book" WHERE "modeladmintest_book"."author_id" = 2 ORDER BY "modeladmintest_book"."id" ASC LIMIT 1
33. SELECT "modeladmintest_book"."id", "modeladmintest_book"."author_id", "modeladmintest_book"."title", "modeladmintest_book"."cover_image_id" FROM "modeladmintest_book" WHERE "modeladmintest_book"."author_id" = 2 ORDER BY "modeladmintest_book"."id" DESC LIMIT 1
34. SELECT "modeladmintest_book"."id", "modeladmintest_book"."author_id", "modeladmintest_book"."title", "modeladmintest_book"."cover_image_id" FROM "modeladmintest_book" WHERE "modeladmintest_book"."author_id" = 1 ORDER BY "modeladmintest_book"."id" ASC LIMIT 1
35. SELECT "modeladmintest_book"."id", "modeladmintest_book"."author_id", "modeladmintest_book"."title", "modeladmintest_book"."cover_image_id" FROM "modeladmintest_book" WHERE "modeladmintest_book"."author_id" = 1 ORDER BY "modeladmintest_book"."id" DESC LIMIT 1
```

The failing CI runs have no `auth_permission` queries and do not have the content type query, maybe there's a difference in the test setup?

```
1. SELECT "django_session"."session_key", "django_session"."session_data", "django_session"."expire_date" FROM "django_session" WHERE ("django_session"."expire_date" > '2021-04-22 13:04:48.890921' AND "django_session"."session_key" = 'cdxt3haywolxk97znfm01gw7vxoj69xx') LIMIT 21
2. SELECT "customuser_customuser"."password", "customuser_customuser"."last_login", "customuser_customuser"."is_superuser", "customuser_customuser"."identifier", "customuser_customuser"."username", "customuser_customuser"."email", "customuser_customuser"."is_staff", "customuser_customuser"."is_active", "customuser_customuser"."first_name", "customuser_customuser"."last_name", "customuser_customuser"."country", "customuser_customuser"."attachment" FROM "customuser_customuser" WHERE "customuser_customuser"."identifier" = 1794430612 LIMIT 21
3. SELECT "wagtailusers_userprofile"."id", "wagtailusers_userprofile"."user_id", "wagtailusers_userprofile"."submitted_notifications", "wagtailusers_userprofile"."approved_notifications", "wagtailusers_userprofile"."rejected_notifications", "wagtailusers_userprofile"."updated_comments_notifications", "wagtailusers_userprofile"."preferred_language", "wagtailusers_userprofile"."current_time_zone", "wagtailusers_userprofile"."avatar" FROM "wagtailusers_userprofile" WHERE "wagtailusers_userprofile"."user_id" = 1794430612 LIMIT 21
4. SELECT COUNT(*) AS "__count" FROM "modeladmintest_author"
5. SELECT COUNT(*) AS "__count" FROM "modeladmintest_author"
6. SELECT COUNT(*) AS "__count" FROM "modeladmintest_author"
7. SELECT "wagtailcore_locale"."id", "wagtailcore_locale"."language_code" FROM "wagtailcore_locale" ORDER BY "wagtailcore_locale"."language_code" ASC
8. SELECT (1) AS "a" FROM "wagtailcore_page" WHERE "wagtailcore_page"."content_type_id" IN (28, 65, 83, 81, 75, 59) LIMIT 1
9. SELECT (1) AS "a" FROM "wagtailcore_page" LIMIT 1
10. SELECT (1) AS "a" FROM "wagtailcore_page" WHERE "wagtailcore_page"."depth" = 1 LIMIT 1
11. SELECT "wagtailcore_page"."path" FROM "wagtailcore_page" WHERE "wagtailcore_page"."depth" = 1
12. SELECT "wagtailcore_page"."id", "wagtailcore_page"."path", "wagtailcore_page"."depth", "wagtailcore_page"."numchild", "wagtailcore_page"."translation_key", "wagtailcore_page"."locale_id", "wagtailcore_page"."title", "wagtailcore_page"."draft_title", "wagtailcore_page"."slug", "wagtailcore_page"."content_type_id", "wagtailcore_page"."live", "wagtailcore_page"."has_unpublished_changes", "wagtailcore_page"."url_path", "wagtailcore_page"."owner_id", "wagtailcore_page"."seo_title", "wagtailcore_page"."show_in_menus", "wagtailcore_page"."search_description", "wagtailcore_page"."go_live_at", "wagtailcore_page"."expire_at", "wagtailcore_page"."expired", "wagtailcore_page"."locked", "wagtailcore_page"."locked_at", "wagtailcore_page"."locked_by_id", "wagtailcore_page"."first_published_at", "wagtailcore_page"."last_published_at", "wagtailcore_page"."latest_revision_created_at", "wagtailcore_page"."live_revision_id", "wagtailcore_page"."alias_of_id" FROM "wagtailcore_page" WHERE "wagtailcore_page"."path" = '0001' LIMIT 21
13. SELECT (1) AS "a" FROM "wagtailcore_page" LIMIT 1
14. SELECT (1) AS "a" FROM "wagtailcore_page" LIMIT 1
15. SELECT "modeladmintest_author"."id", "modeladmintest_author"."name", "modeladmintest_author"."date_of_birth" FROM "modeladmintest_author" ORDER BY "modeladmintest_author"."id" DESC LIMIT 5
16. SELECT "modeladmintest_book"."id", "modeladmintest_book"."author_id", "modeladmintest_book"."title", "modeladmintest_book"."cover_image_id" FROM "modeladmintest_book" WHERE "modeladmintest_book"."author_id" = 5 ORDER BY "modeladmintest_book"."id" ASC LIMIT 1
17. SELECT "modeladmintest_book"."id", "modeladmintest_book"."author_id", "modeladmintest_book"."title", "modeladmintest_book"."cover_image_id" FROM "modeladmintest_book" WHERE "modeladmintest_book"."author_id" = 5 ORDER BY "modeladmintest_book"."id" DESC LIMIT 1
18. SELECT "modeladmintest_book"."id", "modeladmintest_book"."author_id", "modeladmintest_book"."title", "modeladmintest_book"."cover_image_id" FROM "modeladmintest_book" WHERE "modeladmintest_book"."author_id" = 4 ORDER BY "modeladmintest_book"."id" ASC LIMIT 1
19. SELECT "modeladmintest_book"."id", "modeladmintest_book"."author_id", "modeladmintest_book"."title", "modeladmintest_book"."cover_image_id" FROM "modeladmintest_book" WHERE "modeladmintest_book"."author_id" = 4 ORDER BY "modeladmintest_book"."id" DESC LIMIT 1
20. SELECT "modeladmintest_book"."id", "modeladmintest_book"."author_id", "modeladmintest_book"."title", "modeladmintest_book"."cover_image_id" FROM "modeladmintest_book" WHERE "modeladmintest_book"."author_id" = 3 ORDER BY "modeladmintest_book"."id" ASC LIMIT 1
21. SELECT "modeladmintest_book"."id", "modeladmintest_book"."author_id", "modeladmintest_book"."title", "modeladmintest_book"."cover_image_id" FROM "modeladmintest_book" WHERE "modeladmintest_book"."author_id" = 3 ORDER BY "modeladmintest_book"."id" DESC LIMIT 1
22. SELECT "modeladmintest_book"."id", "modeladmintest_book"."author_id", "modeladmintest_book"."title", "modeladmintest_book"."cover_image_id" FROM "modeladmintest_book" WHERE "modeladmintest_book"."author_id" = 2 ORDER BY "modeladmintest_book"."id" ASC LIMIT 1
23. SELECT "modeladmintest_book"."id", "modeladmintest_book"."author_id", "modeladmintest_book"."title", "modeladmintest_book"."cover_image_id" FROM "modeladmintest_book" WHERE "modeladmintest_book"."author_id" = 2 ORDER BY "modeladmintest_book"."id" DESC LIMIT 1
24. SELECT "modeladmintest_book"."id", "modeladmintest_book"."author_id", "modeladmintest_book"."title", "modeladmintest_book"."cover_image_id" FROM "modeladmintest_book" WHERE "modeladmintest_book"."author_id" = 1 ORDER BY "modeladmintest_book"."id" ASC LIMIT 1
25. SELECT "modeladmintest_book"."id", "modeladmintest_book"."author_id", "modeladmintest_book"."title", "modeladmintest_book"."cover_image_id" FROM "modeladmintest_book" WHERE "modeladmintest_book"."author_id" = 1 ORDER BY "modeladmintest_book"."id" DESC LIMIT 1
```

Reply:
@frmdstryr Have you changed any configuration on your side at all? E.g. using `DummyCache` backend instead of something that actually caches in some way could be a cause of something like this.

On the content type query.... Try importing the `ContentType` model from `django.contrib.contenttypes.models` adding the following before your  `Site.get_site_root_paths()` line:
```
ContentType.clear_cache()
```

This should level the playing field in across setups.

Reply:
@frmdstryr Also, are you running this test in isolation locally? This can often make a difference.

As you can see from [this line](https://github.com/wagtail/wagtail/blob/main/wagtail/contrib/modeladmin/options.py#L125) the permission helper is instantiated at the time of registering the ModelAdmin class, and that same `ModelAdmin` instance and `PermissionHelper` instance persists in memory throughout the duration of the tests.

So, if there are multiple tests present that load the listing page, the test that 'runs first' will trigger the query and cache the result, THEN, the next test that loads the list will benefit from the cache.

Reply:
So locally an env with ES always does 35 queries irregardless if running the single test or all tests, but an env without ES does 35 if running the single test and only 25 when running all tests.